### PR TITLE
*: remove kube-rbac-proxy sidecar container

### PR DESCRIPTION
**Pull request by simonpasquier**
This commit removes the need of running a kube-rbac-proxy sidecar
container next to the CMO container by implementing TLS +
authentication/authorization directly in CMO.

Supported authentication methods are client TLS certificates and bearer
tokens. Client TLS certificates are verified when issued by the
Kubernetes cluster's CA (read from the 'client-ca-file' key in the
kube-system/extension-apiserver-authentication ConfigMap).

There's a built-in authorization rule for the
'system:serviceaccount:openshift-monitoring:prometheus-k8s' user
(service account) on the '/metrics' endpoint. Other users will be
allowed based on their RBAC permissions.

The following endpoints are also available without authn/authz:
* '/healthz'
* '/livez'
* '/readyz'

The implementation mostly builds upon the
github.com/openshift/library-go and k8s.io/apiserver modules. This is
why many of the Go dependencies have been updated.

* [ ] I added CHANGELOG entry for this change.
* [X] No user facing changes, so no entry in CHANGELOG was needed.

### CODE REVIEW

**Review by jan--f — CHANGES_REQUESTED** [review]
Very cool, thank you!
This looks great, I only have one nit and a request to drop the `WIP` from the commit message. While you're add it maybe also mention that this updates a whole bunh of dependencies?

**Review by JoaoBraveCoding — APPROVED** [review]
Awesome work! I would agree with Jan on improving the commit message!
So a TL;DR for this PR is:
- Replacing kube-rbac-proxy with a dedicated server to serve CMO metrics
- Adhering to the k8s lib for metrics, which allows us to set metric stabilityLevel

**Comment by simonpasquier** [comment]
> This looks great, I only have one nit and a request to drop the WIP from the commit message.

@jan--f thanks for noticing!

**Comment by simonpasquier** [comment]
updated the commit message and the PR description :)

File: assets/cluster-monitoring-operator/service-monitor.yaml [hunk 1]
```diff
@@ -7,11 +7,19 @@ metadata:
   namespace: openshift-monitoring
 spec:
   endpoints:
-  - bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
+  - bearerTokenFile: ""
+    metricRelabelings:
+    - action: drop
```
Comment (raptorsun) [inline assets/cluster-monitoring-operator/service-monitor.yaml:12]: why these metrics are dropped?

Comment (jan--f) [reply assets/cluster-monitoring-operator/service-monitor.yaml:12]: The [jsonnet code](https://github.com/openshift/cluster-monitoring-operator/pull/1870/files#diff-a89d64ee6e92f3dd1e7faf2c0e4bdf29ea537f3a82d7dd8db62490a0e8799efc) has the answer :)

**Review by jan--f — APPROVED** [review]
/lgtm
:+1: :ship:

**Comment by simonpasquier** [comment]
/label px-approved
/label docs-approved

No user facing changes so adding the PX and docs labels myself. I'll let @juzhao and/or @lihongyan1 verify that we have no regression and add the `qe-approved` label.

**Comment by juzhao** [comment]
tested with PR, cluster-monitoring-operator endpoint exposes more metrics than nightly payload now, example: (apiserver|go_sched|workqueue)_.+ metrics, and the metrics are dropped from prometheus, according to the PR, this is expected:
PR metrics: http://pastebin.test.redhat.com/1089772
nightly payload metrics: http://pastebin.test.redhat.com/1089770

/lgtm

**Pull request merged**